src/common/socket.h

```diff
@@ -303,5 +290,16 @@ extern int32 naddr_;   // # of ip addresses
 	int32 sendudp(int32 fd,void *buff,size_t nbytes,int32 flags,const struct sockaddr *from,socklen_t addrlen);
 #endif 
 
+template<typename T, typename U>
+T& ref(U* buf, std::size_t index)
+{
+    return *reinterpret_cast<T*>(buf + index);
+}
+
+template<typename T, typename U = void*>
```

Comment:
what is this?

Reply:
```cpp
void* thing;
thing + 1; // error, void* unknown size
```

If I don't provide the overload, it means adding more (uint8*) into message, login_auth, etc.  Where we are using zmq and get data() (void*).